inspector: break in eval script

eugeneo opened the pull request:
Fixes: https://github.com/nodejs/node/issues/14577

<!--
Thank you for your pull request. Please provide a description above and review
the requirements below.

Bug fixes and new features should include tests and possibly benchmarks.

Contributors guide: https://github.com/nodejs/node/blob/master/CONTRIBUTING.md
-->

##### Checklist
<!-- Remove items that do not apply. For completed items, change [ ] to [x]. -->

- [x] `make -j4 test` (UNIX), or `vcbuild test` (Windows) passes
- [x] tests and/or benchmarks are included
- [x] commit message follows [commit guidelines](https://github.com/nodejs/node/blob/master/CONTRIBUTING.md#commit-message-guidelines)

##### Affected core subsystem(s)
eval scripts are now wrapped into inspector call to break on a first line.

eugeneo commented:
What I don't like about this solution is that the wrapper is reported to frontend. E.g. running `console.log(3)` shows the following source:

```javascript
process.binding('inspector').callAndPauseOnStart(function() {

console.log(3);

}, {})
```
Alternative would be to modify the node_contextify.cc to add a parameter to break on the first line of the code ran in the context. I am not sure this is something that should be done for just debugging the eval scripts.

addaleax commented on lib/internal/bootstrap_node.js:
```diff
@@ -420,7 +427,7 @@
     module.paths = Module._nodeModulePaths(cwd);
-    const body = process._eval;
+    const body = wrapForBreakOnFirstLine(process._eval);
     const script = `global.__filename = ${JSON.stringify(name)};\n` +
```
would it work better if you applied `wrapForBreakOnFirstLine` to `script`, not to `body`?

eugeneo replied:
Then the entire block will be reported as source and actual user code will be a string - e.g. it will not be possible to set a breakpoint there. Stepping into the vm module will be required in order to step into the user code.

BridgeAR commented:
PTAL @nodejs/v8-inspector

BridgeAR commented:
@nodejs/collaborators PTAL

addaleax commented:
LGTM. I feel like there could be a few weird edge cases with this approach, but if so I guess just seeing when they show up is fine.

CI: https://ci.nodejs.org/job/node-test-commit/12414/

addaleax commented:
@eugeneo I think this might need to be updated

eugeneo commented:
I added "in progress" label as this PR needs some rework due to changes elsewhere.

eugeneo commented:
As it turns out, a number of tests were relying on this bug. I updated them to expect the new breakpoint and adjusted the line numbers.

CI: https://ci.nodejs.org/job/node-test-pull-request/10127/

addaleax commented:
@eugeneo by the way, the typo is still in the commit message … ;)

eugeneo commented:
Another CI attempt - https://ci.nodejs.org/job/node-test-pull-request/10146/

joaocgreis commented:
Windows CI failed because of an issue that's already fixed. Here is another run with the same parameters: https://ci.nodejs.org/job/node-test-commit-windows-fanned/11803/

eugeneo commented:
This CI does not seems to indicate any regressions caused by this PR: https://ci.nodejs.org/job/node-test-pull-request/10147/

eugeneo commented:
Landed as https://github.com/nodejs/node/commit/75606c4f6901ce4086addb0e54b1e88d108eaf8b
